src/commands/sgx/runner.rs

```diff
@@ -0,0 +1,103 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+use aesm_client::AesmClient;
+use bytes::{Buf, BufMut};
+use enclave_runner::usercalls::{SyncStream, UsercallExtension};
+use enclave_runner::EnclaveBuilder;
+use sgxs_loaders::isgx::Device as IsgxDevice;
+use std::cell::RefCell;
+use std::io::Cursor;
+use std::io::Result as IoResult;
+use std::io::{Read, Write};
+use std::ops::DerefMut;
+use std::path::Path;
+
+/// User call extension allow the enclave code to "connect" to an external service via a customized enclave runner.
+/// Here we customize the runner to intercept calls to connect to an address "cat" which actually connects the enclave application to
+/// stdin and stdout of `cat` process.
+
+pub struct SgxServer;
+
+thread_local! {
+    pub static BUFFER: RefCell<Cursor<Vec<u8>>> = RefCell::new(Cursor::new(vec![]));
+}
+
+pub fn read_from_buffer(dest: &mut [u8]) -> IoResult<usize> {
+    BUFFER.with(|cell| {
+        let mut cursor = cell.borrow_mut();
+        let mut reader = cursor.deref_mut().reader();
+        reader.read(dest)
+    })
+}
+
+pub fn write_to_buffer(src: &[u8]) -> IoResult<usize> {
+    BUFFER.with(|cell| {
+        let mut cursor = cell.borrow_mut();
+        let mut writer = cursor.get_mut().writer();
+        writer.write(src)
+    })
+}
+
+impl Read for SgxServer {
+    fn read(&mut self, buf: &mut [u8]) -> IoResult<usize> {
+        read_from_buffer(buf)
+    }
+}
+
+impl SyncStream for SgxServer {
+    fn read(&self, dest: &mut [u8]) -> IoResult<usize> {
+        read_from_buffer(dest)
+    }
+
+    fn write(&self, src: &[u8]) -> IoResult<usize> {
+        write_to_buffer(src)
+    }
+
+    fn flush(&self) -> IoResult<()> {
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct ExternalService;
+// Ignoring local_addr and peer_addr, as they are not relavent in the current context.
+impl UsercallExtension for ExternalService {
+    fn connect_stream(
+        &self,
+        addr: &str,
+        _local_addr: Option<&mut String>,
+        _peer_addr: Option<&mut String>,
+    ) -> IoResult<Option<Box<dyn SyncStream>>> {
+        // If the passed address is not "sgx", we return none, whereby the passed address gets treated as
+        // an IP address which is the default behavior.
+        match &*addr {
+            "sgx" => {
+                let stream = SgxServer;
+                Ok(Some(Box::new(stream)))
+            }
+            _ => Ok(None),
+        }
+    }
+}
+
+pub fn run_sgx<P: AsRef<Path>>(file: P) {
+    let mut device = IsgxDevice::new()
+        .expect("get sgx device failed")
+        .einittoken_provider(AesmClient::new())
+        .build();
+    let mut enclave_builder = EnclaveBuilder::new(file.as_ref());
+    enclave_builder
+        .coresident_signature()
+        .expect("sign enclave failed");
+    enclave_builder.usercall_extension(ExternalService);
+    let enclave = enclave_builder
+        .build(&mut device)
+        .expect("get enclave failed");
```

Comment:
Does this need to be executed for every command? Wouldn't it make sense to do it first and then just do ` enclave.run()`?
Ot `main` in enclave code could keep reading the stream infinitely, so `enclave.run()` could be started on a separate thread (this would probably need a lazy_static and mutex on the buffer)?

Reply:
I think I have to do like this. Actually, I can't put the `infinite stream` in the enclave app code, if i did so, I have to hardcode a `ip:port` in the enclave app code. And after I put the `infinite stream` in the runner,  for each comming request, I have to run `enclave.run()`, and `run()` here will consume the `enclave` which defined as `fn run(self)`, so I have to create a new enclave for each very command.

Reply:
`enclave.run()` will consume the `enclave` where `run` defined as `def run(self)`, and the enclave app will run the `main()` function.  If we want the enclave `main()` function keep reading the stream infinitely, it will act as a server, then I have to hardcode a `ip:port` in the enclave app main function, which is not expected. 

So, I have to make the enclave app act like a client, which `main` function will run and exit. For each comming request, I have to run the `enclave.run()`, and the `enclave` should be a new one.

And I think it's Ok here, because `sgx` cli command is just run and the process will exit. You can't run two command at same time in just one process.

Reply:
what about `tmkms start` command? that's the main one that will keep running

Reply:
`tmkms start` is only related to the `sgx server`,  which is the `signatory-sgx/bin/server`. I have already fixed it.